fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,431 @@
+```
+
+#### `is_ffg_confirmed`
+
+```python
+def is_ffg_confirmed(
+    # Returns whether the branch will justify it's current epoch checkpoint at the end of this epoch
+    store: Store,
+    confirmation_byzantine_threshold: int,
+    confirmation_slashing_threshold: int,
+    block_root: Root,
+) -> bool:
+    current_slot = get_current_slot(store)
+    block = store.blocks[block_root]
+    assert get_current_store_epoch(store) == compute_epoch_at_slot(block.slot)
+
+    current_epoch = get_current_store_epoch(store)
+    checkpoint_root = get_checkpoint_block(store, block_root, current_epoch)
+    checkpoint_state = store.block_states[checkpoint_root]
+
+    remaining_ffg_weight = int(get_remaining_weight_in_epoch(store, current_slot))
+    total_active_balance = int(get_total_active_balance(checkpoint_state))
+    current_weight_in_epoch = total_active_balance - remaining_ffg_weight
+    assert current_weight_in_epoch >= 0
+
+    ffg_support_for_checkpoint = int(get_ffg_support(store, block_root))
+    max_adversarial_ffg_support_for_checkpoint = int(
+        min(
+            (current_weight_in_epoch * confirmation_byzantine_threshold - 1) // 100 + 1,
+            confirmation_slashing_threshold,
+            ffg_support_for_checkpoint
+        )
+    )
+
+    return (
+        2 / 3 * total_active_balance <= (
+            ffg_support_for_checkpoint - max_adversarial_ffg_support_for_checkpoint +
+            (1 - confirmation_byzantine_threshold / 100) * remaining_ffg_weight
+        )
+    )
+```
+
+### `is_confirmed`
+
+```python
+def is_confirmed(
+    store: Store,
+    confirmation_byzantine_threshold: int,
+    confirmation_slashing_threshold: int,
+    block_root: Root
+) -> bool:
+    current_epoch = get_current_store_epoch(store)
+
+    block = store.blocks[block_root]
+    block_state = store.block_states[block_root]
+    block_justified_checkpoint = block_state.current_justified_checkpoint.epoch
+
+    # This function is only applicable to current epoch blocks
+    assert compute_epoch_at_slot(block.slot) == current_epoch
+
+    return (
+        block_justified_checkpoint + 1 == current_epoch
+        and is_lmd_confirmed(store, confirmation_byzantine_threshold, block_root)
+        and is_ffg_confirmed(store, confirmation_byzantine_threshold, confirmation_slashing_threshold, block_root)
+    )
+```
+
+## Safe Block Hash
+
+This function is used to compute the value of the `safeBlockHash` field which is passed from CL to EL in the `forkchoiceUpdated` Engine API call.
+
+### Helper Functions
+
+#### `find_confirmed_block`
+
+```python
+def find_confirmed_block(
+    store: Store,
+    confirmation_byzantine_threshold: int,
+    confirmation_slashing_threshold: int,
+    block_root: Root
+) -> Root:
+
+    block = store.blocks[block_root]
+    current_epoch = get_current_store_epoch(store)
+
+    if compute_epoch_at_slot(block.slot) != current_epoch:
+        return store.finalized_checkpoint.root
```

Comment:
I am not sure that `justified checkpoint` is necessarily confirmed.

Reply:
It's weird to introduce a new concept that is weaker than the old one in some cases. Would appreciate if it could be researched. IMO, justification requires a supermajority to attest to a block (and, implicitly, all of its parents)

Reply:
And, does this mean that we should do intermediate versions that report `finalized` as safe block, instead of `justified`? Because, if `justified` is not necessarily confirmed, I'm not sure if reporting it as safe to the EL is correct. Current safe block spec reports justified though.

Reply:
I think that `justified` can become unsafe today under very extreme conditions; and with a new algorithm `justified` can be unconfirmed only under similar conditions. So, I would not say that the new weakens what we have